### PR TITLE
Backports 0.15 pr3

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
will rebase after pr2 is merged

### CODE REVIEW

**Comment by UdjinM6** [comment]
Needs rebase (#2597 was just merged)

**Comment by PastaPastaPasta** [comment]
updated

**Comment by UdjinM6** [comment]
Smth probably went wrong here cause I see commits from pr2 now. Needs proper rebase again.

File: src/test/miner_tests.cpp [hunk 1]
```diff
@@ -249,7 +258,7 @@ BOOST_AUTO_TEST_CASE(CreateNewBlock_validity)
     }
 
     // Just to make sure we can still make simple blocks
-    BOOST_CHECK(pblocktemplate = BlockAssembler(chainparams).CreateNewBlock(scriptPubKey));
+    BOOST_CHECK(pblocktemplate = AssemblerForTest(chainparams).CreateNewBlock(scriptPubKey));
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/test/miner_tests.cpp]: https://github.com/dashpay/dash/pull/2609/commits/51476994043fa245e02cf575d892cbffa1c0de46

File: src/miner.cpp [hunk 2]
```diff
@@ -81,24 +81,38 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
...
+    blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
+    nBlockMaxSize = DEFAULT_BLOCK_MAX_SIZE;
+}
+
+BlockAssembler::BlockAssembler(const CChainParams& params, const Options& options) : chainparams(params)
 {
-    // Largest block you're willing to create:
-    nBlockMaxSize = GetArg("-blockmaxsize", DEFAULT_BLOCK_MAX_SIZE);
+    blockMinFeeRate = options.blockMinFeeRate;
+    // Limit size to between 1K and MAX_BLOCK_SERIALIZED_SIZE-1K for sanity:
```
Comment (UdjinM6) [inline src/miner.cpp]:
```suggestion
    // Limit size to between 1K and MaxBlockSize()-1K for sanity:
```

File: src/wallet/test/wallet_tests.cpp [hunk 3]
```diff
@@ -488,4 +488,29 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
...
+
+    // Call GetImmatureCredit() once before adding the key to the wallet to
+    // cache the current immature credit amount, which is 0.
+    BOOST_CHECK_EQUAL(wtx.GetImmatureCredit(), 0);
+
+    // Invalidate the cached value, add the key, and make sure a new immature
+    // credit amount is calculated.
+    wtx.MarkDirty();
+    wallet.AddKeyPubKey(coinbaseKey, coinbaseKey.GetPubKey());
+    BOOST_CHECK_EQUAL(wtx.GetImmatureCredit(), 50*COIN);
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/wallet/test/wallet_tests.cpp]: I remember doing this already... weird https://github.com/dashpay/dash/pull/2609/commits/1fa7f7e743b250659904e16b905d1d68af2bd7cd

**Pull request merged**